### PR TITLE
Allow `cluster_networking_config` to have `defaults` origin

### LINKED ISSUE #19099: `cluster_networking_config` with default values has `teleport.dev/origin: config-file` label instead of `teleport.dev/origin: defaults`
### Expected behavior:

If the `auth_service` configuration has no `cluster_networking_config` fields, the `cluster_networking_config` should have the `teleport.dev/origin: defaults` label.
 
### Current behavior:

If the `auth_service` configuration has no `cluster_networking_config` fields, the `cluster_networking_config` has the `teleport.dev/origin: config-file` label.

As noted in https://github.com/gravitational/teleport/issues/18829#issuecomment-1337500610, this seems to be due to this line: https://github.com/gravitational/teleport/blob/8bdcd197f1951fadafd777959bbb518ccc65bafc/lib/config/configuration.go#L661

### Bug details:
- Teleport version: v11.1.1
- Recreation steps:
```bash
teleport configure -o file
teleport start
tctl get cluster_networking_config
```

### PR DESCRIPTION
**Pull request by vitorenesduarte**
Fixes #19099.

This commit ensures that `cluster_networking_config` has the `teleport.dev/origin: defaults` label when no related field is set in the configuration file.
Before this commit, `cluster_networking_config` would always have the `teleport.dev/origin: config-file`, even if no related field was set.

### CODE REVIEW

File: lib/config/configuration.go [hunk 1]
```diff
@@ -635,6 +635,7 @@ func applyAuthConfig(fc *FileConfig, cfg *service.Config) error {
 	}
 
 	if fc.Auth.MessageOfTheDay != "" {
+		cfg.Auth.Preference.SetOrigin(types.OriginConfigFile)
```
Comment (vitorenesduarte) [inline lib/config/configuration.go:638]: I believe this was missed in de9bdf086dd5471278ddc7ab1fb6d9a66c9f9882.

File: lib/config/configuration.go [hunk 2]
```diff
@@ -661,26 +662,40 @@ func applyAuthConfig(fc *FileConfig, cfg *service.Config) error {
...
+	customNetworkingConfig := fc.Auth.ClientIdleTimeout != 0 ||
+		fc.Auth.ClientIdleTimeoutMessage != "" ||
+		fc.Auth.WebIdleTimeout != 0 ||
+		fc.Auth.KeepAliveInterval != 0 ||
+		fc.Auth.KeepAliveCountMax != 0 ||
+		fc.Auth.SessionControlTimeout != 0 ||
+		fc.Auth.ProxyListenerMode != 0 ||
+		fc.Auth.RoutingStrategy != 0 ||
+		fc.Auth.TunnelStrategy != nil ||
+		fc.Auth.ProxyPingInterval != 0
```
Comment (marcoandredinis) [inline lib/config/configuration.go:676]: :nail_care: 
Can we extract this check into a function?
And compare each field of the `fc.Auth` against a default `config.Auth` (as in `x := config.Auth{}`)?

**Review by marcoandredinis — APPROVED** [review]

File: lib/config/configuration.go (hunk 2 above)
Comment (vitorenesduarte) [reply lib/config/configuration.go]: Thanks for the suggestion! Done in 857fd1e0ea06d03ba157d2b627f1e758600df414.

File: lib/config/configuration.go (hunk 1 above)
Comment (nklaassen) [reply lib/config/configuration.go:638]: What if the origin is already `OriginDynamic`? `MessageOfTheDay` seems less consequential than the rest of the auth preferences, I'm not sure it makes sense to overwrite the origin for this, but I don't know much about what these origin tags are used for

File: lib/config/configuration.go [hunk 3]
```diff
@@ -661,26 +662,29 @@ func applyAuthConfig(fc *FileConfig, cfg *service.Config) error {
...
-		KeepAliveCountMax:        fc.Auth.KeepAliveCountMax,
-		SessionControlTimeout:    fc.Auth.SessionControlTimeout,
-		ProxyListenerMode:        fc.Auth.ProxyListenerMode,
-		RoutingStrategy:          fc.Auth.RoutingStrategy,
-		TunnelStrategy:           fc.Auth.TunnelStrategy,
-		ProxyPingInterval:        fc.Auth.ProxyPingInterval,
-	})
-	if err != nil {
-		return trace.Wrap(err)
+	// Only override networking configuration if some of its fields is
```
Comment (nklaassen) [inline lib/config/configuration.go:665]: grammar nit:
```suggestion
	// Only override networking configuration if some of its fields are
```

**Review by nklaassen — APPROVED** [review]

File: lib/config/configuration_test.go [hunk 4]
```diff
@@ -778,6 +778,9 @@ func TestApplyConfig(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, types.AgentMesh, tunnelStrategyType)
 	require.Equal(t, types.DefaultAgentMeshTunnelStrategy(), cfg.Auth.NetworkingConfig.GetAgentMeshTunnelStrategy())
+	require.Equal(t, cfg.Auth.Preference.Origin(), types.OriginConfigFile)
```
Comment (zmb3) [inline lib/config/configuration_test.go:781]: These assertions are backwards. The expected value should come before the actual value.

File: lib/config/fileconf.go [hunk 5]
```diff
@@ -889,6 +889,30 @@ type Auth struct {
 	LoadAllCAs bool `yaml:"load_all_cas,omitempty"`
 }
 
+// hasCustomNetworkingConfig returns true if any of the networking
+// configuration fields have values different from an empty Auth.
+func (a *Auth) hasCustomNetworkingConfig() bool {
+	empty := Auth{}
+	return a.ClientIdleTimeout != empty.ClientIdleTimeout ||
```
Comment (zmb3) [inline lib/config/fileconf.go:896]: Why not just compare these to the zero value without creating the empty auth struct first?

File: lib/config/configuration_test.go (hunk 4 above)
Comment (vitorenesduarte) [reply lib/config/configuration_test.go]: Thanks! Done in 3a7053dd7864b6193f43cb4201d0af7eaef03381.

**Comment by vitorenesduarte** [comment]
> @nklaassen thanks for raising this. This PR as is can indeed change the outcome of `shouldInitReplaceResourceWithOrigin`:
> 
> https://github.com/gravitational/teleport/blob/4f89756d8b8391d8c696d2c07652dc9afc6215b1/lib/auth/init.go#L527-L552
> 
> This can result in a breaking change (in a seemingly unlikely situation) as shown below.
>
> If a given configuration only has the `message_of_the_day` `cluster_auth_preference` field, then:
> 
> * before this PR, it would be a candidate with `defaults` (which seems wrong, and that's why I added this line)
> * with this PR, it will be a candidate with `config-file`
> 
> Now, if there's already a `cluster_auth_preference` that is `dynamic` (as you asked), then:
> 
> * before this PR we had situation `#7`, and so the `dynamic` resource is kept
> * with this PR we have situation `#8`, and so the `dynamic` resource is discarded with a warning
> 
> This means that this PR would be a breaking change. 

@nklaassen @zmb3 Given the above, I decided to revert this small change in dfbad2d0a573958811d10521c2fd8f84bdddb496 so that this PR is just about #19099.

dfbad2d0a573958811d10521c2fd8f84bdddb496 also adds a test case describing the current behaviour and I opened #19272 in case we want to fix this afterwards.

**Pull request merged**